### PR TITLE
Fixes #13983 - Environment can be inserted into Environments Path

### PR DESCRIPTION
**Pull request by xprazak2**

### CODE REVIEW

File: app/models/katello/kt_environment.rb [hunk 1]
```diff
@@ -102,6 +100,30 @@ def successor
       self.organization.promotion_paths[0][0] unless self.organization.promotion_paths.empty?
     end
 
+    # creates new env from create_params with self as a prior
+    def insert_successor(create_params, path)
+      binding.pry
```
Comment (daviddavis) [inline app/models/katello/kt_environment.rb]: Hello.

Comment (stbenjam) [inline app/models/katello/kt_environment.rb]: Hi :wave:

File: test/models/kt_environment_test.rb [hunk 2]
```diff
@@ -78,5 +78,11 @@ def test_content_view_label_excludes_content_dir
       # this an ActiveModel::Errors object; not a Hash
       assert env.errors.include?(:label)
     end
+
+    def test_insert_successor_after_library
+       @library.insert_successor({ :organization => @acme_corporation, :name => "testEnv" }, @dev.path)
+       assert @library.successors.map(&:name).include?("testEnv")
+       assert_equal "testEnv", @dev.prior.name
+    end
```
Comment (daviddavis) [inline test/models/kt_environment_test.rb]: I'd probably roll a couple more tests for cases like library, etc.

File: app/models/katello/kt_environment.rb [hunk 3]
```diff
@@ -102,6 +100,29 @@ def successor
...
+        new_successor.save
+      else
+        old_successor = successor
+        new_successor.prior = self
+        old_successor.prior = new_successor
+        new_successor.save
+      end
+      fail HttpErrors::UnprocessableEntity, _('An environment is missing a prior') unless all_have_prior?
+      new_successor
+    end
```
Comment (daviddavis) [inline app/models/katello/kt_environment.rb]: Two things on this method. First can you refactor out `new_successor.prior = self` and `new_successor.save`? Seems like you duplicate that logic a couple times. Also, I'm wondering if we could/should wrap this method in a transaction.

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/environments/views/new-environment.html [hunk 4]
```diff
@@ -56,6 +52,18 @@ <h1 translate>New Environment</h1>
         </textarea>
       </div>
 
+
+
```
Comment (daviddavis) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/environments/views/new-environment.html]: nitpick: you probably don't need these two blank lines.

**Comment by beav** [comment]
needs rebase, setting waiting on contributor

**Comment by beav** [comment]
@xprazak2 is this PR still ok? I can test after rebase if so.

**Comment by xprazak2** [comment]
Rebased, it should be ok now.

**Comment by beav** [comment]
[test katello]

File: db/migrate/20160302091113_change_environment_prior.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class ChangeEnvironmentPrior < ActiveRecord::Migration
```
Comment (beav) [inline db/migrate/20160302091113_change_environment_prior.rb]: this needs a `[4.2]` at the end (or whichever version you choose I suppose)

Comment (beav) [reply db/migrate/20160302091113_change_environment_prior.rb]: this should get the katello test further along as well

**Comment by beav** [comment]
I was able to create an environment path and add an env to the middle using the "prior" field. I also was able to promote content through (yay).

I created "dev" and "prod", then created a second path of "yy" and "zz", and added "xx" to the second path. However, when I edited the description for environment xx, I got "An error occurred saving the Environment: environment can only have one child".

You may also want to rename "prior" to "prior environment"in the web UI, just to be clearer.

**Comment by beav** [comment]
@xprazak2 Thanks for updating, the bug no longer exists :beetle:. When I tried this in the web UI everything worked as expected.

It looks like there is one rubocop failure and some unit test failures. The change looks good other than that.

**Comment by xprazak2** [comment]
[test katello]

**Comment by beav** [comment]
[test katello]

previous failure appeared to be unrelated to PR, rekicking test

**Comment by xprazak2** [comment]
Tests are green now, any additional comments?

**Comment by beav** [comment]
@xprazak2 automation passes, and my manual testing passes (including tests that failed last time). The code also looks good (thanks @daviddavis for earlier review).

ACK, thanks @xprazak2 ! this is a cool feature

**Pull request merged**